File: src/Detect_new_connections.py
""" 
    Script that monitors system calls: inet_create, inet6_create,
    tcp_v4_connect, tcp_v6_connect, inet_dgram_connect, inet_csk_accept,
    udp_sendmsg, udpv6_sendmsg, udp_recvmsg, udpv6_recvmsg. 
    Script creates a log file in logs folder, that contains new network 
    connections, that are anotated by its aplication, or by URL address
    of request in case the connections is made by browser.


    Licensed under the Apache License, Version 2.0 (the "License")
"""

__author__ = "Branislav Dubec"
__version__ = "1.0.5"


from bcc import BPF
import psutil
import os
import threading
from concurrent.futures import ThreadPoolExecutor
import json
import csv
from struct import pack
from socket import inet_ntop, AF_INET, AF_INET6
from urllib.parse import urlparse
from collections import OrderedDict

class LimitedSizeDict(OrderedDict):
    def __init__(self, max_size=1000, file_path=""):
        super().__init__()
        self.max_size = max_size
        self.file_path = file_path

    def __setitem__(self, key, value):
        if key in self:
            self[key].append(value)
            super().move_to_end(key)
        else:
            if len(self) == self.max_size:
                key, items = self.popitem(last=False)
                for (url, line) in items:
                    with open(self.file_path, "a") as no_match_file:
                        no_match_file.write(line + "\n")
                super().__setitem__(key, [value])
                self.move_to_end(key)
            else:
                super().__setitem__(key, [value])
                self.move_to_end(key)
    
    def print_values(self):
        with open(self.file_path, "a") as no_match_file:
            for key,values in self.items():
                for value in values:
                    if type(value) is list:
                        for list_value in value:
                            try:
                                no_match_file.write(list_value[1] + "\n")
                            except:
                                no_match_file.write(str(list_value))
                    else:
                        no_match_file.write(value[1] + "\n")

    def get_pop_item(self,key):
        if key in self:
            values = self[key]
            url, line = values.pop(0)
            if not values:
                del self[key]
            else:
                self[key] = values
            return url
        return ""

### Global variables

# BPF code
b = BPF(src_file="Detect_new_connections.c")

"""
type_dict = {
    1: "SOCK_STREAM",    # TCP
    2: "SOCK_DGRAM",     # UDP
    3: "SOCK_RAW",   
    4: "SOCK_RDM",
    5: "SOCK_SEQPACKET",
    6: "SOCK_DCCP",
    10: "SOCK_PACKET",
}

family_dict = {
    1: "AF_UNIX/AF_LOCAL",  
    2: "AF_INET",           # IPv4 
    10: "AF_INET6",         # IPv6 
    17: "AF_PACKET",       
}

state_dict = {

    1: "TCP_ESTABLISHED",  
    2: "TCP_SYN_SENT",     
    3: "TCP_SYN_RECV",    
    4: "TCP_FIN_WAIT1",    
    5: "TCP_FIN_WAIT2",    
    6: "TCP_TIME_WAIT",    
    7: "TCP_CLOSE",
    8: "TCP_CLOSE_WAIT",
    9: "TCP_LAST_ACK",
    10: "TCP_LISTEN",
    11: "TCP_CLOSING",
    12: "TCP_NEW_SYN_RECV",
}
"""
protocol_dict = {
    0: "IP",
    1: "ICMP",
    6: "TCP",
    17: "UDP",
    41: "IPV6",
    92: "MTP",
    51: "AH",
    143: "ETH",
    255: "RAW",
    262: "MPTCP",
}

functions_to_events = {
    "inet_create": ("new_ipv4_socket_entry", "new_ipv4_socket_return"),
    "inet6_create": ("new_ipv6_socket_entry", "new_ipv6_socket_return"),
    "tcp_v4_connect": ("tcp_connect_entry", "tcp_connect_return"),
    "tcp_v6_connect": ("tcp_connect_entry", "tcp_connect_return"),
    "inet_dgram_connect": ("connect_udp_entry", "connect_udp_return"),
    "inet_csk_accept": ("accept_entry", "accept_return"),
    "udp_sendmsg": ("udp_sendmsg_entry", "udp_sendmsg_return"),
    "udpv6_sendmsg": ("udp_sendmsg_entry", "udp_sendmsg_return"),
    "udp_recvmsg": ("udp_recvmsg_entry", "udp_recvmsg_return"),
    "udpv6_recvmsg": ("udp_recvmsg_entry", "udp_recvmsg_return"),
}


filter_labels = ["systemd-resolve"]
log_file_path = "logs/socket_log.csv"
no_match_file_path = "logs/no_match.json"
pipe_path = "/tmp/firefox_url_pipe"

headers = [
    "PID",
    "TGID",
    "COMMAND",
    "SOURCE ADDRESS",
    "DESTINATION ADDRESS",
    "SOURCE PORT",
    "DESTINATION PORT",
    "LABEL",
    "PROTOCOL",
]

url_ip_browser_data = {}
browser_request_data = LimitedSizeDict(max_size = 1000, file_path="../logs/no_match.json")
known_address = {}

# lock to acquire writing privileges to output file
lock = threading.Lock()
# Thread that waits for messages in mkfifo file
# and is responsible for logging connections created
# by web browsers
def read_from_pipe():

    # set permissions
    old_umask = os.umask(0)

    if not os.path.exists(pipe_path):
        os.mkfifo(pipe_path, 0o666)

    os.umask(old_umask)

    with open(pipe_path, "r") as pipe:

        while True:
            message = pipe.readline().strip()
            if message:
                try:
                    line = json.loads(message)
                except json.JSONDecodeError as e:
                    print(f"Error decoding JSON: {e}", file=sys.stderr)
                    continue
                parseResult = urlparse(line["url"])
                url = parseResult.netloc
                ip = line["ip"]
                port = parseResult.scheme
                if port == "https" or port == "wss":
                    port = 443
                else:
                    port = 80
                hash_value = hash((ip, port))
                key = (ip, port)
                known_address.setdefault(key, set()).add(url)
                if key in url_ip_browser_data:
                    events_list = url_ip_browser_data[key]

                    first_event = events_list.pop(0)

                    if not events_list:
                        del url_ip_browser_data[key]
                    else:
                        url_ip_browser_data[key] = events_list
                    first_event["label"] = first_event["label"] + "-" + url
                    row = list(first_event.values())
                    
                    lock.acquire()
                    with open(log_file_path, "a") as log_file:
                        writer = csv.writer(log_file)
                        writer.writerow(row)
                    lock.release()
                else:
                    browser_request_data[(ip,port)] = (url,json.dumps(line))                                             
                


pipe_thread = threading.Thread(target=read_from_pipe, daemon=True)
pipe_thread.start()
thread_executor = ThreadPoolExecutor(max_workers=5)


def log_event(event):
    comm = event.comm.decode("UTF-8")
    if filter_event(comm):
        return
    try:
        src_ip = inet_ntop(AF_INET, pack("I", event.saddr))
        dst_ip = inet_ntop(AF_INET, pack("I", event.daddr))
    except ValueError as e:
        src_ip = inet_ntop(AF_INET6, event.saddr)
        dst_ip = inet_ntop(AF_INET6, event.daddr)

    try:
        label = get_application_name(event.pid)
    except Exception as e:
        label = "No Process found"
    used_protocol = protocol_dict[event.protocol]
    pid = event.pid
    tgid = event.tgid
    dport = event.dport
    sport = event.sport
    row = [pid, tgid, comm, src_ip, dst_ip, sport, dport, label, used_protocol]

    if is_new_web_request(comm, label, dst_ip):
        key = (dst_ip, dport)
        url = browser_request_data.get_pop_item(key)
        if url:
            label = label + "-" + url
            row = [pid, tgid, comm, src_ip, dst_ip, sport, dport, label, used_protocol]
            lock.acquire()
            with open(log_file_path, "a") as log_file:
                writer = csv.writer(log_file)
                writer.writerow(row)
            lock.release()
            return
        # hash_ip_port = hash((dstip,event.dport))
        row_dict = {
            "pid": pid,
            "tgid": tgid,
            "comm": comm,
            "srcip": src_ip,
            "dstip": dst_ip,
            "sport": sport,
            "dport": dport,
            "label": label,
            "protocol": used_protocol,
        }

        url_ip_browser_data.setdefault(key, []).append(row_dict)
        # url_ip_browser_data[hash_ip_port] = event
    else:
        lock.acquire()
        with open(log_file_path, "a") as log_file:
            writer = csv.writer(log_file)
            writer.writerow(row)
        lock.release()


def attach_function_to_event(event_name, entry_function, return_function):
    global b
    event_name_bytes = bytes(event_name, "utf-8")
    if b.get_kprobe_functions(event_name_bytes):
        b.attach_kprobe(event=event_name, fn_name=entry_function)
        b.attach_kretprobe(event=event_name, fn_name=return_function)
    else:
        print(
            f"ERROR: {event_name} kernel not found or traceable."
            f"The kernel might be too old or the function has been inlined."
        )


def get_application_name(pid):
    try:
        process = psutil.Process(pid)
        return process.name()
    except Exception as e:
        print(f"Not found process for PID: %d", pid)
        return ""


# Filter events, DNS Res or systemd-resolved
def filter_event(comm):
    return comm in filter_labels or comm.startswith("DNS Res")


def is_new_web_request(comm, label, dstip):
    if comm == "Socket Thread":
        if label == "firefox":  # add different browsers
            return True
    return False


def print_ipv6_event(cpu, data, size):
    event = b["ipv6events"].event(data)

    thread_executor.submit(log_event, event)


def print_ipv4_event(cpu, data, size):
    event = b["ipv4events"].event(data)

    thread_executor.submit(log_event, event)


if __name__ == "__main__":

    for event, (entry_function, return_function) in functions_to_events.items():
        attach_function_to_event(event, entry_function, return_function)

    lock.acquire()
    with open(log_file_path, "a") as log_file:
        writer = csv.writer(log_file)
        writer.writerow(headers)
    lock.release()

    b["ipv4events"].open_perf_buffer(print_ipv4_event)
    b["ipv6events"].open_perf_buffer(print_ipv6_event)

    print("Logging new socket connections. Press Ctrl+C to stop.")
    try:
        while True:
            b.perf_buffer_poll()

    except KeyboardInterrupt:
        browser_request_data.print_values()
        lock.acquire()
        with open(log_file_path, "a") as log_file:
            writer = csv.writer(log_file)
            for key, values in url_ip_browser_data.items():
                for value in values:
                    if key in known_address:
                        url_set = known_address[key]
                        if len(url_set) == 1:
                            url = next(iter(url_set))
                        #else:
                            #url = ''.join(url_set)
                    else:
                        url = ""
                    value["label"] = value["label"] + "-" + str(url)
                    
                    writer.writerow(value.values())
        lock.release()
        print("Stopped logging socket connections.")

File: src/Detect_new_connections.c
/*
    Contains definition of functions that are connected to system calls.
    Functions ending with entry is executed before the function.
    Functions ending with return is execture after the function.

    Licensed under the Apache License, Version 2.0 (the "License")

    __author__ = Branislav Dubec
    __version__ = 1.0.5

*/

#include "Detect_new_connections.h"

/*
 Current version does not track lifetime of sockets, therefore no events are produced to user mode.
*/
// Creation of inet_socket,
// Updates eBPF map with pointer of current socket with its pid_tgid
int new_ipv4_socket_entry(struct pt_regs *ctx, struct net *net, struct socket *sock) {

    u64 pid_tgid = bpf_get_current_pid_tgid();  

    new_socket.update(&pid_tgid, &sock);
    
    return 0;
}

// Return of creation of inet_socket
int new_ipv4_socket_return(struct pt_regs *ctx) {
    struct data_ipv4 data = {};

    int ret = PT_REGS_RC(ctx);
    
    u64 pid_tgid = bpf_get_current_pid_tgid();

    struct socket **socketpp;
    socketpp = new_socket.lookup(&pid_tgid);

    if (socketpp == NULL) {
    bpf_trace_printk("missed_entry in ipv4 socket creation");
        return 0;   // missed entry
    }

    if (ret != 0) {
        bpf_trace_printk("error return create ipv4 socket %d", ret);
        new_socket.delete(&pid_tgid);
        return 0;
    }
   
    return 0;
}


// Creation of inet6_socket
// Updates eBPF map with pointer of current socket with its pid_tgid
int new_ipv6_socket_entry(struct pt_regs *ctx, struct net *net, struct socket *sock, int protocol, int kern) {

    
    u64 pid_tgid = bpf_get_current_pid_tgid();
    
    new_socket.update(&pid_tgid, &sock);
    
    return 0;
}

// Return of creation of inet_socket -> function is done and returns
int new_ipv6_socket_return(struct pt_regs *ctx) {
    struct data_ipv6 data = {};

    int ret = PT_REGS_RC(ctx);
    
    u64 pid_tgid = bpf_get_current_pid_tgid();

    struct socket **socketpp;
    socketpp = new_socket.lookup(&pid_tgid);

    if (socketpp == NULL) {
        bpf_trace_printk("missed_entry in ipv6 socket creation");
        return 0;   // missed entry
    }

    if (ret != 0) {
        bpf_trace_printk("error return create ipv6 socket %d", ret);
        new_socket.delete(&pid_tgid);
        return 0;
    }    
    return 0;
}



// Beginning of tcp connection for both IPv4 and IPv6
// Updates eBPF map with pointer of current sock with its pid_tgid
int tcp_connect_entry(struct pt_regs *ctx, struct sock *sk)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    new_sock.update(&pid_tgid, &sk);
    return 0;
}

// Return of tcp connection for both IPv4 and IPv6: tcp_v4_connect, tcp_v6_connect
// Check if the return value is success.
// Get information from sock structure and send it to user mode.
int tcp_connect_return(struct pt_regs *ctx)
{

    int ret = PT_REGS_RC(ctx);

    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tgid = pid_tgid;

    struct sock **skpp;
    skpp = new_sock.lookup(&pid_tgid);
    if (skpp == NULL) {
        bpf_trace_printk("Error: missed entry in tcp socket connect return.");
        return 0;   // missed entry
    }
    if (ret != 0) {
        bpf_trace_printk("Error: tcp connect return value: %d.", ret);
        new_sock.delete(&pid_tgid);
        return 0;
    }

    struct sock *skp = *skpp;
    u16 lport = skp->__sk_common.skc_num;
    u16 dport = skp->__sk_common.skc_dport;
    u16 protocol = skp->sk_protocol;
    u16 family = skp->sk_family;
    if (family == AF_INET){
        struct data_ipv4 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.protocol = protocol;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        data.dport = ntohs(dport);
        data.pid = pid;
        data.tgid = tgid;
        data.saddr = skp->__sk_common.skc_rcv_saddr;
        data.daddr = skp->__sk_common.skc_daddr;
        ipv4events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        return 0;
    } else { // family is AF_INET6
        struct data_ipv6 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        data.dport = ntohs(dport);
        data.pid = pid;
        data.tgid = tgid;
        bpf_probe_read_kernel(&data.saddr, sizeof(data.saddr),
                           &skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr),
                           &skp->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
        ipv6events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        return 0;
    }

}

// Beginning of tcp accept function for both IPv4 and IPv6, server connection, inet_csk_accept function
// Updates eBPF map with pointer of current sock with its pid_tgid
int accept_entry(struct pt_regs *ctx, struct sock *sk, int flags, int *err)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    new_sock.update(&pid_tgid, &sk);
    return 0;
}

// Return of tcp accept function for both IPv4 and IPv6
// If unsuccessful, function returns NULL
// Get information from sock structure and send it to user mode.
int accept_return(struct pt_regs *ctx)
{


    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tgid = pid_tgid;

    

    struct sock **skpp;
    skpp = new_sock.lookup(&pid_tgid);
    if (skpp == NULL) {
        bpf_trace_printk("Error: missed entry in tcp socket accept return.");
        return 0;   // missed entry
    }
    
    struct sock *newsk = (struct sock *)PT_REGS_RC(ctx);

    if (newsk == NULL){
        bpf_trace_printk("Error: tcp connect return value NULL.");
        new_sock.delete(&pid_tgid);
        return 0;
    }
    
    

    u16 lport = newsk->__sk_common.skc_num;
    u16 dport = newsk->__sk_common.skc_dport;
    u16 protocol = newsk->sk_protocol;
    u16 family = newsk->sk_family;
    if (family == AF_INET){
        struct data_ipv4 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.protocol = protocol;
        data.type = newsk->sk_type;
        data.state = newsk->sk_state;
        data.sport = lport;
        data.dport = ntohs(dport);
        data.pid = pid;
        data.tgid = tgid;
        data.saddr = newsk->__sk_common.skc_rcv_saddr;
        data.daddr = newsk->__sk_common.skc_daddr;
        ipv4events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        return 0;
    } else {
        struct data_ipv6 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.type = newsk->sk_type;
        data.state = newsk->sk_state;
        data.sport = lport;
        data.dport = ntohs(dport);
        data.pid = pid;
        data.tgid = tgid;
        bpf_probe_read_kernel(&data.saddr, sizeof(data.saddr),
                           &newsk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr),
                           &newsk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
        ipv6events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        return 0;
    }

}


// Beginning of tcp accept function for both IPv4 and IPv6 inet_dgram_connect function
// Updates eBPF map with pointer of current sock with its pid_tgid
int connect_udp_entry(struct pt_regs *ctx, struct socket *sock, struct sockaddr *uaddr,
		       int addr_len, int flags)
{

    u64 pid_tgid = bpf_get_current_pid_tgid();
    new_socket.update(&pid_tgid, &sock);
   
    return 0;
}


// Return of tcp accept function for both IPv4 and IPv6
// Check if the return value is successfull
// Get information from sock structure and send it to user mode.
int connect_udp_return(struct pt_regs *ctx)
{
    

    int ret = PT_REGS_RC(ctx);
    
    u64 pid_tgid = bpf_get_current_pid_tgid();

    struct socket **socketpp;
    socketpp = new_socket.lookup(&pid_tgid);

    if (socketpp == NULL) {
        bpf_trace_printk("Error: missed entry in udp connect return.");
        return 0;   // missed entry
    }
    if (ret != 0) {
        bpf_trace_printk("Error: return connect upd msg %d.", ret);
        new_socket.delete(&pid_tgid);
        return 0;
    }

    struct socket *socketp = *socketpp;

    u32 pid = pid_tgid >> 32;
    u32 tgid = pid_tgid;

    u16 family = socketp->sk->sk_family;

    if (family == AF_INET) {
        struct data_ipv4 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.type = socketp->sk->sk_type;
        data.state = socketp->sk->sk_state;
        data.saddr = socketp->sk->__sk_common.skc_rcv_saddr;
        data.daddr = socketp->sk->__sk_common.skc_daddr;
        data.pid = pid;
        data.tgid = tgid;
        data.sport = socketp->sk->__sk_common.skc_num;
        data.dport = socketp->sk->__sk_common.skc_dport;
        data.dport = ntohs(data.dport);
        ipv4events.perf_submit(ctx, &data, sizeof(data));
        new_socket.delete(&pid_tgid);
        return 0;
    }
    else {
        struct data_ipv6 data = {};
        data.pid = pid;
        data.tgid = tgid;
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.type = socketp->sk->sk_type;
        data.state = socketp->sk->sk_state;
        bpf_probe_read_kernel(&data.saddr, sizeof(data.saddr),
            socketp->sk->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr),
            socketp->sk->__sk_common.skc_v6_daddr.in6_u.u6_addr32);
        data.sport = socketp->sk->__sk_common.skc_num;
        data.dport = socketp->sk->__sk_common.skc_dport;
        data.dport = ntohs(data.dport);
        ipv6events.perf_submit(ctx, &data, sizeof(data));
        new_socket.delete(&pid_tgid);
        return 0;
    }
}

// Beginning of UDP send msg, for both IPv4 and IPv6, udp_sendmsg and udpv6_sendmsg functions
// Updates eBPF maps with sock and msg -> to get destination
int udp_sendmsg_entry(struct pt_regs *ctx, struct sock *sk, struct msghdr *msg, size_t len)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    new_sock.update(&pid_tgid, &sk);  
    new_msg.update(&pid_tgid, &msg); 
    return 0;
}

// Return of udp send msg function for both IPv4 and IPv6
// Check if the return value < 0 that means that no msg was sent
// Check if sock contains information about destination -> 
//          destination was therefore already anotated via connect_udp
// Get information from sock structure and from msg structure and send it to user mode.
int udp_sendmsg_return(struct pt_regs *ctx)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tgid = pid_tgid;

    int ret = PT_REGS_RC(ctx);


    struct sock **skpp;
    skpp = new_sock.lookup(&pid_tgid);
    if (skpp == NULL) {
        bpf_trace_printk("Error: missed entry in udp sendmsg return.");
        return 0;   // missed entry
    }
    struct msghdr **msgp;
    msgp = new_msg.lookup(&pid_tgid);
    if (msgp == NULL) {
        bpf_trace_printk("Error: missed entry in udp sendmsg return.");
        return 0;   // missed entry
    }
    if (ret < 0){
        bpf_trace_printk("Error return udp sendmsg  %d.", ret);
        new_sock.delete(&pid_tgid);
        new_msg.delete(&pid_tgid);
        return 0;
    }
    struct sock *skp = *skpp;
    
    
    u16 family = skp->sk_family;
    u16 protocol = skp->sk_protocol;
    u16 lport = skp->__sk_common.skc_num;
    struct msghdr *msg = *msgp;
    if (family == AF_INET) {//IPv4
        if (skp->__sk_common.skc_daddr != 0){ // connection is traced viac udp_connect

            bpf_trace_printk("UDP sendmsg known address %d", skp->__sk_common.skc_daddr);
            new_sock.delete(&pid_tgid);
            new_msg.delete(&pid_tgid);
            return 0;
        }
        struct sockaddr_in *usin = (struct sockaddr_in *)msg->msg_name;

        struct data_ipv4 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.protocol = protocol;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        
        data.pid = pid;
        data.tgid = tgid;
        data.saddr = skp->__sk_common.skc_rcv_saddr;

        data.daddr = usin->sin_addr.s_addr;
        u16 dport = usin->sin_port;
        data.dport = ntohs(dport);
        ipv4events.perf_submit(ctx, &data, sizeof(data));
        return 0;
    }
    else{
        struct sockaddr_in6 *usin6 = (struct sockaddr_in6 *)msg->msg_name;
        struct data_ipv6 data = {};
        bpf_probe_read_kernel(&data.saddr, sizeof(data.saddr),
                           &skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr), 
                           &skp->__sk_common.skc_v6_daddr.in6_u.u6_addr32);

        if (data.daddr != 0 || !usin6){ // connection is traced viac udp_connect

            bpf_trace_printk("UPDv6 sendmsg known address %d.", skp->__sk_common.skc_daddr);
            new_sock.delete(&pid_tgid);
            new_msg.delete(&pid_tgid);
            return 0;
        }

        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.protocol = protocol;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        
        data.pid = pid;
        data.tgid = tgid;
        
        data.dport = usin6->sin6_port;
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr), 
                           &usin6->sin6_addr);

        ipv6events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        new_msg.delete(&pid_tgid);
        return 0;

    }
        
}

// Beginning of UDP recv msg, for both IPv4 and IPv6, udp_recvmsg and udpv6_recvmsg functions
// Updates eBPF maps with sock and msg -> to get destination
int udp_recvmsg_entry(struct pt_regs *ctx, struct sock *sk, struct msghdr *msg, 
    size_t len,int flags, int *addr_len)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    new_sock.update(&pid_tgid, &sk);  
    new_msg.update(&pid_tgid, &msg); 
    return 0;
}

// Return of udp recv msg function for both IPv4 and IPv6
// Check if the return value < 0 that means that no msg was sent
// Check if sock contains information about destination -> 
//          destination was therefore already anotated via connect_udp
// Get information from sock structure and from msg structure and send it to user mode.
int udp_recvmsg_return(struct pt_regs *ctx)
{
    u64 pid_tgid = bpf_get_current_pid_tgid();
    u32 pid = pid_tgid >> 32;
    u32 tgid = pid_tgid;

    int ret = PT_REGS_RC(ctx);


    struct sock **skpp;
    skpp = new_sock.lookup(&pid_tgid);
    if (skpp == NULL) {
        bpf_trace_printk("Error: missed entry in udp sendmsg return.");
        return 0;   // missed entry
    }
    struct msghdr **msgp;
    msgp = new_msg.lookup(&pid_tgid);
    if (msgp == NULL) {
        bpf_trace_printk("Error: missed entry in udp sendmsg return.");
        return 0;   // missed entry
    }

    if (ret < 0){
        bpf_trace_printk("Error return recvmsg udp %d. ", ret);
        new_sock.delete(&pid_tgid);
        new_msg.delete(&pid_tgid);
        return 0;
    }

    struct sock *skp = *skpp;
    u16 family = skp->sk_family;
    u16 protocol = skp->sk_protocol;
    u16 lport = skp->__sk_common.skc_num;
    struct msghdr *msg = *msgp;

    if (family == AF_INET) {//IPv4
        if (skp->__sk_common.skc_daddr != 0){ // connection is traced viac udp_connect

            bpf_trace_printk("Udp recvmsg known address %d.", skp->__sk_common.skc_daddr);
            new_sock.delete(&pid_tgid);
            new_msg.delete(&pid_tgid);
            return 0;
        }
        struct sockaddr_in *usin = (struct sockaddr_in *)msg->msg_name;

        struct data_ipv4 data = {};
        bpf_get_current_comm(data.comm, sizeof(data.comm));
        
        data.family = family;
        data.protocol = protocol;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        
        data.pid = pid;
        data.tgid = tgid;
        data.saddr = skp->__sk_common.skc_rcv_saddr;

        data.daddr = usin->sin_addr.s_addr;
        u16 dport = usin->sin_port;
        data.dport = ntohs(dport);
        ipv4events.perf_submit(ctx, &data, sizeof(data));
        return 0;
    }
    else{
        struct sockaddr_in6 *usin6 = (struct sockaddr_in6 *)msg->msg_name;
        struct data_ipv6 data = {};
        bpf_probe_read_kernel(&data.saddr, sizeof(data.saddr),
                           &skp->__sk_common.skc_v6_rcv_saddr.in6_u.u6_addr32);
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr), 
                           &skp->__sk_common.skc_v6_daddr.in6_u.u6_addr32);

        if (data.daddr != 0 || !usin6){ // connection is traced viac udp_connect

            bpf_trace_printk("UDPv6 recvmsg known address %d.", skp->__sk_common.skc_daddr);
            new_sock.delete(&pid_tgid);
            new_msg.delete(&pid_tgid);
            return 0;
        }

        bpf_get_current_comm(data.comm, sizeof(data.comm));
        data.family = family;
        data.protocol = protocol;
        data.type = skp->sk_type;
        data.state = skp->sk_state;
        data.sport = lport;
        
        data.pid = pid;
        data.tgid = tgid;
        
        data.dport = usin6->sin6_port;
        bpf_probe_read_kernel(&data.daddr, sizeof(data.daddr), 
                           &usin6->sin6_addr);

        ipv6events.perf_submit(ctx, &data, sizeof(data));
        new_sock.delete(&pid_tgid);
        new_msg.delete(&pid_tgid);
        return 0;

    }
}